File: src/ec2/metadata.rs
use crate::errors::{
    Error::{Other, API},
    Result,
};
use chrono::{DateTime, Utc};
use reqwest::ClientBuilder;
use serde::{Deserialize, Serialize};
use tokio::time::Duration;

/// Fetches the instance ID on the host EC2 machine.
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
pub async fn fetch_instance_id() -> Result<String> {
    fetch_metadata_by_path("instance-id").await
}

/// Fetches the public hostname of the host EC2 machine.
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
pub async fn fetch_public_hostname() -> Result<String> {
    fetch_metadata_by_path("public-hostname").await
}

/// Fetches the public IPv4 address of the host EC2 machine.
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
pub async fn fetch_public_ipv4() -> Result<String> {
    fetch_metadata_by_path("public-ipv4").await
}

/// Fetches the availability of the host EC2 machine.
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
pub async fn fetch_availability_zone() -> Result<String> {
    fetch_metadata_by_path("placement/availability-zone").await
}

/// Fetches the spot instance action.
///
/// If Amazon EC2 is not stopping or terminating the instance, or if you terminated the instance yourself,
/// spot/instance-action is not present in the instance metadata thus returning an HTTP 404 error.
///
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/prepare-for-interruptions.html
pub async fn fetch_spot_instance_action() -> Result<InstanceAction> {
    let resp = fetch_metadata_by_path("spot/instance-action").await?;
    serde_json::from_slice(resp.as_bytes()).map_err(|e| Other {
        message: format!(
            "failed to parse spot/instance-action response '{}' {:?}",
            resp, e
        ),
        is_retryable: false,
    })
}

/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-instance-termination-notices.html#instance-action-metadata
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub struct InstanceAction {
    pub action: String,
    #[serde(with = "rfc_manager::serde_format::rfc_3339")]
    pub time: DateTime<Utc>,
}

/// Fetches the region of the host EC2 machine.
/// TODO: fix this...
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
pub async fn fetch_region() -> Result<String> {
    let mut az = fetch_availability_zone().await?;
    az.truncate(az.len() - 1);
    Ok(az)
}

/// Fetches instance metadata service v2 with the "path".
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
/// e.g., curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/public-ipv4
pub async fn fetch_metadata_by_path(path: &str) -> Result<String> {
    log::info!("fetching meta-data/{}", path);

    let token = fetch_token().await?;

    let uri = format!("http://169.254.169.254/latest/meta-data/{}", path);
    let cli = ClientBuilder::new()
        .user_agent(env!("CARGO_PKG_NAME"))
        .danger_accept_invalid_certs(true)
        .timeout(Duration::from_secs(15))
        .connection_verbose(true)
        .build()
        .map_err(|e| API {
            message: format!("failed ClientBuilder build {:?}", e),
            is_retryable: false,
        })?;
    let resp = cli
        .get(&uri)
        .header("X-aws-ec2-metadata-token", token)
        .send()
        .await
        .map_err(|e| API {
            message: format!("failed to build GET meta-data/{} {:?}", path, e),
            is_retryable: false,
        })?;
    let out = resp.bytes().await.map_err(|e| API {
        message: format!("failed to read bytes {:?}", e),
        is_retryable: false,
    })?;
    let out: Vec<u8> = out.into();

    match String::from_utf8(out) {
        Ok(text) => Ok(text),
        Err(e) => Err(API {
            message: format!("GET meta-data/{} failed String::from_utf8 ({})", path, e),
            is_retryable: false,
        }),
    }
}

/// Serves session token for instance metadata service v2.
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
/// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
/// e.g., curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"
const IMDS_V2_SESSION_TOKEN_URI: &str = "http://169.254.169.254/latest/api/token";

/// Fetches the IMDS v2 token.
async fn fetch_token() -> Result<String> {
    log::info!("fetching IMDS v2 token");

    let cli = ClientBuilder::new()
        .user_agent(env!("CARGO_PKG_NAME"))
        .danger_accept_invalid_certs(true)
        .timeout(Duration::from_secs(15))
        .connection_verbose(true)
        .build()
        .map_err(|e| API {
            message: format!("failed ClientBuilder build {:?}", e),
            is_retryable: false,
        })?;
    let resp = cli
        .put(IMDS_V2_SESSION_TOKEN_URI)
        .header("X-aws-ec2-metadata-token-ttl-seconds", "21600")
        .send()
        .await
        .map_err(|e| API {
            message: format!("failed to build PUT api/token {:?}", e),
            is_retryable: false,
        })?;
    let out = resp.bytes().await.map_err(|e| API {
        message: format!("failed to read bytes {:?}", e),
        is_retryable: false,
    })?;
    let out: Vec<u8> = out.into();

    match String::from_utf8(out) {
        Ok(text) => Ok(text),
        Err(e) => Err(API {
            message: format!("GET token failed String::from_utf8 ({})", e),
            is_retryable: false,
        }),
    }
}

File: aws/go/ec2/metadata/metadata.go
package metadata

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/gyuho/infra/aws/go/pkg/logutil"
)

// Serves session token for instance metadata service v2.
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
// e.g., curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"
const IMDS_V2_SESSION_TOKEN_URI = "http://169.254.169.254/latest/api/token"

// Fetches the IMDS v2 token.
func FetchToken(ctx context.Context) (string, error) {
	req, err := http.NewRequest(http.MethodPut, IMDS_V2_SESSION_TOKEN_URI, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", "21600")

	cli := &http.Client{}
	resp, err := cli.Do(req.WithContext(ctx))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// Fetches instance metadata service v2 with the "path".
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
// e.g., curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/public-ipv4
func FetchMetadataByPath(ctx context.Context, path string) (string, error) {
	uri := fmt.Sprintf("http://169.254.169.254/latest/meta-data/%s", path)
	logutil.S().Infow("fetching meta-data", "uri", uri)

	token, err := FetchToken(ctx)
	if err != nil {
		return "", err
	}
	req, err := http.NewRequest(http.MethodGet, uri, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("X-aws-ec2-metadata-token", token)

	cli := &http.Client{}
	resp, err := cli.Do(req.WithContext(ctx))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// Fetches the instance ID on the host EC2 machine.
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
func FetchInstanceID(ctx context.Context) (string, error) {
	return FetchMetadataByPath(ctx, "instance-id")
}

// Fetches the public hostname of the host EC2 machine.
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
func FetchPublicHostname(ctx context.Context) (string, error) {
	return FetchMetadataByPath(ctx, "public-hostname")
}

// Fetches the public IPv4 address of the host EC2 machine.
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
func FetchPublicIPV4(ctx context.Context) (string, error) {
	return FetchMetadataByPath(ctx, "public-ipv4")
}

// Fetches the availability of the host EC2 machine.
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
func FetchAvailabilityZone(ctx context.Context) (string, error) {
	return FetchMetadataByPath(ctx, "placement/availability-zone")
}

// Fetches the region of the host EC2 machine.
// TODO: fix this...
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
func FetchRegion(ctx context.Context) (string, error) {
	az, err := FetchAvailabilityZone(ctx)
	if err != nil {
		return "", err
	}
	return az[:len(az)-1], nil
}

// Represents the instance action.
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-instance-termination-notices.html#instance-action-metadata
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
type InstanceAction struct {
	Action string `json:"action"`

	// Time in RFC339 format in UTC.
	Time time.Time `json:"time"`
}

// Fetches the spot instance action.
//
// If Amazon EC2 is not stopping or terminating the instance, or if you terminated the instance yourself,
// spot/instance-action is not present in the instance metadata thus returning an HTTP 404 error.
//
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
// ref. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/prepare-for-interruptions.html
func FetchSpotInstanceAction(ctx context.Context) (InstanceAction, error) {
	s, err := FetchMetadataByPath(ctx, "spot/instance-action")
	if err != nil {
		return InstanceAction{}, err
	}
	action := InstanceAction{}
	if err := json.Unmarshal([]byte(s), &action); err != nil {
		return InstanceAction{}, err
	}
	return action, nil
}
